DRA: improve resource slice publishing

pohly opened the pull request:
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

This fixes support for dropped fields when partitionable device fields are used and not supported (see first commit).

More importantly, drivers can now be in informed about dropped fields and invalid slices (see second commit).


#### Does this PR introduce a user-facing change?
```release-note
NONE
```

/cc @mortent

pohly committed:
DRA resourceslices: avoid update loop when partitionable devices feature is off

To emulate the apiserver behavior with optional features disabled, we need to
drop fields in the reactors. When fields get dropped, we want to be sure that
the controller does not try to updated again. The unit test covers that now by
calling ctrl.run again.

The test case for partitionable devices got copy-and-pasted without actually
updating it.

This new testing confirms that the controller needs to be updated to handle
also dropping of spec-level fields. Only dropping of fields in devices was
handled.

pohly committed:
DRA resourceslices: better error reporting

A user of the controller can register an error handler via the controller
options. For a kubelet plugin, the error handler is method in the interface
which must be implemented. This is a conscious choice to make DRA driver
developer aware that they should react intelligently to errors.

The controller will invoke that handler with all errors that it encounters
while syncing the desired set of slices. This includes validation errors from
the apiserver if the driver's slices are invalid. Dropped fields get reported
with a special DroppedFieldsError.

pohly commented on test/integration/dra/dra_test.go:
```diff
@@ -135,9 +136,33 @@ func TestDRA(t *testing.T) {
+				})
+			},
+		},
+		// "v1beta2": {
```
There's currently a commented out v1beta2 in the integration tests because I intend to continue with this PR by adding transparent usage of v1beta1 and v1beta2 by drivers.

jackfrancis commented on staging/src/k8s.io/dynamic-resource-allocation/resourceslice/resourceslicecontroller.go:
```diff
@@ -213,8 +215,72 @@ type Options struct {
+	var disabled []string
+
+	// Both slices should have the same number of devices, but better check it.
+	for i := 0; i < len(err.DesiredSlice.Spec.Devices) && i < len(err.ActualSlice.Spec.Devices); i++ {
```
just to be clear we're not really doing anything if one slice or the other is greater than (we would simply not enumerate through any devices for the type that has the greater number), is that what we want?

pohly replied:
There's nothing else that we can do: we need two of everything. But as the comment says, we should never get into a situation where storing a slice removes devices.

jackfrancis replied:
Is the "need two of everything" requirement upstream from this method as well? Should this be enforced somewhere in `(c *Controller) syncPool` if so?

If we know that the lengths of `err.DesiredSlice.Spec.Devices` and `err.ActualSlice.Spec.Devices` are equal, then we can simplify to

`for i := 0; i < len(err.DesiredSlice.Spec.Devices) {`

which is a nano-optimization and maybe a little easier on the eyes

consider this a nit, not a big deal

jackfrancis commented on staging/src/k8s.io/dynamic-resource-allocation/resourceslice/resourceslicecontroller.go:
```diff
@@ -284,24 +369,26 @@ func newController(ctx context.Context, options Options) (*Controller, error) {
-	for poolName := range options.Resources.Pools {
-		c.queue.Add(poolName)
-	}
+	c.Update(options.Resources)
```
Now that we're calling Update() as part of the constructor function we're going to call it twice in a row in `staging/src/k8s.io/dynamic-resource-allocation/kubeletplugin/draplugin.go` via `(d *Helper) PublishResources` when our `d.resourceSliceController` is nil and we need to go through the flow of constructing a new resource slice controller. Is that okay?

pohly replied:
https://github.com/kubernetes/kubernetes/blob/908bdb3f2c487061975b9a06ca0480e3388bc285/staging/src/k8s.io/dynamic-resource-allocation/kubeletplugin/draplugin.go#L629
ensures that the `PublishResources` is not called when a new controller got constructed, so it won't be done twice.

But this Go pattern of avoiding `else` clauses to reduce indention isn't really doing us any favor here. I've changed it.

jackfrancis replied:
You're right, I missed the early return! I do think your change is more expressive of the flow, thx.

pohly commented on staging/src/k8s.io/dynamic-resource-allocation/resourceslice/resourceslicecontroller.go:
```diff
@@ -233,12 +299,31 @@ func (c *Controller) Update(resources *DriverResources) {
+	// Truncate all timestamps to seconds because that is all that we can store.
+	// Without this we would get semantic differences between desired and actual
+	// stored slice.
+	for _, pool := range c.resources.Pools {
```
I moved it out to simplify the code, but I don't think it's worth adding a dedicated unit test. I had added a test case.

pohly committed:
DRA resourceslices: clarify code

Using an else clause makes it more obvious that the code either creates a new
controller or calls Publish on an existing one, not both.

mortent commented on staging/src/k8s.io/dynamic-resource-allocation/resourceslice/resourceslicecontroller.go:
```diff
@@ -405,19 +495,9 @@ func (c *Controller) processNextWorkItem(ctx context.Context) bool {
 	defer c.queue.Done(poolName)
 	logger := klog.FromContext(ctx)
 
-	// Panics are caught and treated like errors.
```
Seems like we no longer want to catch panics. Any reason why this was needed before, but not needed now?

pohly replied:
I changed my mind: The normal behavior of Kubernetes components is to crash after a panic, because that is something that shouldn't happen and needs to be surfaced more obviously than merely with an error log entry. I now decided to follow that approach.

mortent commented on staging/src/k8s.io/dynamic-resource-allocation/resourceslice/resourceslicecontroller.go:
```diff
@@ -489,214 +582,201 @@ func (c *Controller) syncPool(ctx context.Context, poolName string) error {
+		slice.Spec.Devices = copyTaintTimeAdded(slice.Spec.Devices, pool.Slices[i].Devices)
+
+		logger.V(5).Info("Updating existing resource slice", "slice", klog.KObj(slice))
+		actualSlice, err := c.kubeClient.ResourceV1beta1().ResourceSlices().Update(ctx, slice, metav1.UpdateOptions{})
```
So here we actually do the update and then decide if fields were dropped. This means that the ResourceSlices with the missing fields are in the APIServer and will be used by the scheduler with potentially incorrect results. I remember we discussed trying to use dry-run to detect this situation without actually making the incorrect ResourceSlices live in the cluster. Did we try that out?

pohly replied:
No, I haven't. @johnbelamaric convinced me that plan A should be that admins downgrade drivers before downgrading the cluster. This check is relatively cheap and doesn't add much complexity, but a dry-run isn't, therefore I decided against doing it.

pohly committed:
DRA: v1beta1 <-> v1beta2 conversion

The manual code is an almost literal copy of the pkg/apis/resource/v1beta1
code. The "internal" version here is the latest version of our API, which will
change over time. The intent is to wrote code using that latest API version and
then to convert to older versions automatically.

pohly committed:
DRA: client-go wrapper with latest API as types

The wrapper currently takes and returns v1beta2 types and under the hood
converts to and from whatever type is supported by the apiserver. It tries with
v1beta2 and falls back to v1beta1 only if needed. It then keeps using that for
future calls (i.e. there should only be one failing call). If the apiserver
configuration changes, the wrapper tries the other version again. This is
transparent to the caller.

Patching and server-side-apply are not supported and return an error.

pohly committed:
DRA: use v1beta2

DRA drivers must provide ResourceSlices using the v1beta2 API types.
The controller then converts under the hood to v1beta1 if needed, i.e.
drivers are compatible with Kubernetes 1.32 and Kubernetes 1.33, as
long as at least one beta API group is enabled.

Testing pivots from using v1beta1 as the main API to v1beta2, with only one
test case exercising v1beta1.

bg-chun commented on staging/src/k8s.io/dynamic-resource-allocation/api/v1beta1/conversion.go:
```diff
@@ -0,0 +1,274 @@
+	return nil
+}
+
+func Convert_v1beta2_ResourceSliceSpec_To_v1beta1_ResourceSliceSpec(in *resourceapi.ResourceSliceSpec, out *resourcev1beta1.ResourceSliceSpec, s conversion.Scope) error {
```
Just to clarify, this v1beta2-to-v1beta1 conversion is handled systematically by the API server.  
This is for the case when objects are created using the v1beta2 type, but some older plugins still query using the v1beta1 type.

pohly replied:
The apiserver does a similar conversion to support different clients where one uses v1beta1 and the other uses v1beta2. But it does that under different API endpoints (`/apis/resource.k8s.io/v1beta1/resourceslices` vs `/apis/resource.k8s.io/v1beta2/resourceslices`).

What we want is to let the DRA driver authors use the v1beta2 API, convert on the client side and use the v1beta1 endpoint.

aojea replied:
sorry but I miss context, why do we want to force the use of the v1beta1 endpoint?

pohly replied:
Compatibility with Kubernetes 1.32: that one only supports v1beta1, not v1beta2.

This way, a DRA driver developed with the helper code from v1.33 is compatible with any Kubernetes release >= v1.32.

bg-chun commented on staging/src/k8s.io/dynamic-resource-allocation/api/v1beta1/conversion.go:
```diff
@@ -0,0 +1,274 @@
+	if err := autoConvert_v1beta2_Device_To_v1beta1_Device(in, out, s); err != nil {
+		return err
+	}
+	out.Basic = &resourcev1beta1.BasicDevice{}
```
nit: `BasicDevice` is an optional field in v1beta1.  
I think it makes sense to keep this field as `nil` when the corresponding fields (`Attributes`, `Capacity`, ..., `Taints`) from the v1beta2 object are not set.  These fields are also optional in v1beta2.

pohly replied:
> nit: BasicDevice is an optional field in v1beta1.

No, we decided to only use this one device type also in v1beta1. Even the fields of a "partitionable device" are stored in `BasicDevice`.

nojnhuh commented on staging/src/k8s.io/dynamic-resource-allocation/client/generic.go:
```diff
@@ -0,0 +1,332 @@
+		// all API version until it arrives back at the original one - not
+		// nice because of the overhead, but not wrong semantically
+		// and shouldn't occur often.
+		if apierrors.IsNotFound(err) {
```
Is this how "If the apiserver configuration changes" is detected like the commit message describes, based on hitting a NotFound error?

pohly commented on staging/src/k8s.io/dynamic-resource-allocation/resourceslice/resourceslicecontroller.go:
```diff
@@ -233,18 +299,40 @@ func (c *Controller) Update(resources *DriverResources) {
+	roundTaintTimeAdded(c.resources)
 
 	// ... and the new ones (might be the same).
 	for poolName := range c.resources.Pools {
```
Let's support "nil == no resources" consistently, i.e. not treat it as an error - changed.

pohly commented on staging/src/k8s.io/dynamic-resource-allocation/resourceslice/resourceslicecontroller_test.go:
```diff
@@ -869,6 +961,32 @@ func TestControllerSyncPool(t *testing.T) {
+					if errors.As(err, &droppedFields) {
+						t.Errorf("unexpected dropped fields error: %v\n%s", err, cmp.Diff(droppedFields.DesiredSlice.Spec, droppedFields.ActualSlice.Spec))
+					} else {
+						t.Errorf("unexpected dropped background error: %v", err)
```
No, that was probably copy-and-paste. Applied.

pohly commented on staging/src/k8s.io/dynamic-resource-allocation/resourceslice/resourceslicecontroller_test.go:
```diff
@@ -869,6 +961,32 @@ func TestControllerSyncPool(t *testing.T) {
+			case test.expectedError != "" && len(controllerErrors) == 0:
+				t.Errorf("expected error, got none: %s", test.expectedError)
+			case test.expectedError == "" || test.expectedError != "" && len(controllerErrors) != 1:
+				for _, err := range controllerErrors {
```
I didn't want to make this more complicated than necessary. If no or exactly one error is expected and we got several, then all of them get reported as unexpected. I updated it to emit only one error.

pohly commented on staging/src/k8s.io/dynamic-resource-allocation/api/v1beta1/doc.go:
```diff
@@ -0,0 +1,24 @@
+limitations under the License.
+*/
+
+// k8s:conversion-gen specifies the latest API version.
```
"in k8s.io/api/resource" - added.

pohly commented on staging/src/k8s.io/dynamic-resource-allocation/client/generic.go:
```diff
@@ -0,0 +1,332 @@
+	}
+}
+
+type typedClient[NP objectPtr[N], N, NL, NAC any, OP objectPtr[O], O, OL, OAC any] struct {
```
I renamed to "convertingClient". That it's local is obvious from the code and package description.

pohly commented on staging/src/k8s.io/dynamic-resource-allocation/resourceslice/resourceslicecontroller.go:
```diff
@@ -318,17 +407,43 @@ func (c *Controller) initInformer(ctx context.Context) error {
+	}
+	informer := cache.NewSharedIndexInformer(
+		&cache.ListWatch{
+			ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
```
One should think that I remember that after [adding this](https://github.com/kubernetes/kubernetes/blame/b53b9fb5573323484af9a19cf3f5bfe80760abba/staging/src/k8s.io/client-go/tools/cache/listwatch.go#L152-L163), right? :sweat_smile: 

Fixed.

nojnhuh reviewed (approved):
/lgtm

nojnhuh commented:
> /hold
> 
> This doesn't need to keep being retested until either it's picked for the 1.33 milestone or 1.34 changes are ready to merge.

I see 1.34 PRs are merging now. I'll leave the hold but feel free to remove @pohly if you're satisfied with the reviews here.

pohly commented:
@mortent: you had comments (https://github.com/kubernetes/kubernetes/pull/131246#pullrequestreview-2761508814). If you are satisfied with my responses, then please lift the hold.

pohly commented:
Rebased because of code conflict.

/assign @mortent

pohly commented:
/hold cancel

pohly commented:
There some legitimate linter errors, but they don't show up in https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/131246/pull-kubernetes-linter-hints/1919283244386226176.

Might be related to my recent golangci-lint v2 update.

/test pull-kubernetes-linter-hints

mortent reviewed (approved):
/lgtm
